### PR TITLE
Add support for dynamic log groups using templating variables.

### LINKED ISSUE #6: Intrepolation of the log_group
In may cases the `log_group` is a factor of what this is logging, eg `instances/$SERVICENAME` I wonder if there is a way to interpolate that or make `log_group` get that from a key.

Thanks

### LINKED ISSUE #46: Dynamic Coudwatch Log Groups
Other than shipping it as a sidecar inside every application pod, or building a complex filtering configuration with potentially dozens of instances of the cloudwatch logs output plugin that is dependent on informing every node of every possible application that may or may not be scheduled on it in order to decide ahead of time where logs should go, there appears to be no way to use this plugin to route logs for multiple applications to _separate_ cloudwatch log groups based on the applications.

A perfect solution to my problem would be directly configuring the log group based on kubernetes annotations similar to how they have integrated this feature, https://docs.fluentbit.io/manual/pipeline/filters/kubernetes#annotation-examples-in-pod-definition, but failing that it would be good to have some basic functionality to allow construction of a log_group + log_stream as the tag, and define a separating character/string/index and have the logs routed to multiple log groups based on their "log group prefix" so to speak.

### PR DESCRIPTION
**Pull request by davidnewhall**
Closes #46, closes #6, closes #24, closes #49.

*Description of changes:*

This builds on #78 (more commits to _that_ branch), and can be merged after (making this diff smaller), or in leu of #78.

I create a new struct type called `Event` that carries the Event payload and metadata through the entire request process. Previously the code was passing around the log stream name, and now it passes around the Event which contains the stream name, group name, tag and log record. This makes the data easier to reason about.

- I adapted the same code that created a template from the log stream name to the log group. 
- Added a new `groups` map-tracker to the `OutputPlugin` struct. Gets added to when new groups are created.
- Added the log group name to the `logStream` struct, so each log stream has a pinned unique group.
- Auto create group was removed; groups are now auto-created if they don't exist.
- The group name now supports template variables, explained in the README.

**This config:**
```
    log_group_name      /ecs/twitch-fluentbit/$(tag)/$(host)
    log_stream_name     $(tag[0]).$(ident)
```

**Turns into this (groups):**

Screen Shot 2020-08-19 at 5 34 47 PM
Screen Shot 2020-08-19 at 5 34 35 PM

**Streams Created:**
Screen Shot 2020-08-19 at 5 48 05 PM

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by cwyl02** [comment]
hello when will this get review/merged? This is a feature we are looking for!

**Comment by hossain-rayhan** [comment]
Hey @cwyl02 , sorry its taking longer but we are on it.

File: README.md [hunk 1]
```diff
@@ -23,13 +23,14 @@ Run `make` to build `./bin/cloudwatch.so`. Then use with Fluent Bit:
 ### Plugin Options
 
 * `region`: The AWS region.
-* `log_group_name`: The name of the CloudWatch Log Group that you want log records sent to.
-* `log_stream_name`: The name of the CloudWatch Log Stream that you want log records sent to.
-* `log_stream_prefix`: Prefix for the Log Stream name. The tag is appended to the prefix to construct the full log stream name. Not compatible with the `log_stream_name` option.  
+* `log_group_name`: The name of the CloudWatch Log Group that you want log records sent to. This value allows a template in the form of `${variable}`. See `log_stream_name` description for more. The app will attempt to create missing log groups, and will throw an error if it does not have access.
```
Comment (hossain-rayhan) [inline README.md]: Shouldn't it be `$(variable)`? From #78 conversation, we planned to replace `${ }` with `$( )`. Am I missing something?

File: README.md [hunk 2]
```diff
@@ -23,13 +23,14 @@ Run `make` to build `./bin/cloudwatch.so`. Then use with Fluent Bit:
...
+* `log_stream_name`: The name of the CloudWatch Log Stream that you want log records sent to. This value allows a template in the form of `${variable}` where
+`variable` is a map key name in the log message. To access sub-values in the map
+use the form `${variable['subkey']}`. Special values: `${tag}` references the full tag name, `${tag[0]}` and `${tag[1]}` are the first and second values of log tag split on periods. You may access any member by index, 0 through 9.
```
Comment (hossain-rayhan) [inline README.md]: Same goes for `${tag}`.

File: cloudwatch/cloudwatch.go [hunk 3]
```diff
@@ -211,40 +222,48 @@ func newCloudWatchLogsClient(config OutputPluginConfig) (*cloudwatchlogs.CloudWa
...
-		if err != nil {
-			logrus.Error(err)
-			return fluentbit.FLB_ERROR
-		}
-		output.logGroupCreated = true
-	}
-
-	data, err := output.processRecord(record)
+func (output *OutputPlugin) AddEvent(e *Event) int {
+	// Step 1: convert the Event data to strings, and check for a log key.
```
Comment (hossain-rayhan) [inline cloudwatch/cloudwatch.go:227]: These steps are good. Thanks for making it clear.

File: README.md (hunk 1 above)
Comment (davidnewhall) [reply README.md]: I was waiting for your team to provide feedback on the chosen variable identifiers. If y'all are good with `#()` I'll go ahead and fix the doc. Is there any additional feedback before making a final commit?

Comment (PettitWesley) [reply README.md]: We voted and we prefer `$()`

File: cloudwatch/helpers.go [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+			data = val // drill down another level.
+			return 0, nil
+		default: // missing
+			return w.Write([]byte(tag))
+		}
+	})
+}
+
+// parseDataMapTags parses the provided tag values in template form,
+// from an interface map (expected to contains strings or more maps)
```
Comment (davidnewhall) [inline cloudwatch/helpers.go:66]: typo.

File: README.md (hunk 1 above)
Comment (davidnewhall) [reply README.md]: Sounds good, I'll switch it up and test before committing.

Comment (davidnewhall) [reply README.md]: Added a commit changes the identifiers. Updated the readme. Made some comments better. Fixed a few log lines and added a couple debug lines. If you guys don't mind I'll follow-up with a clean-up PR (no features).

**Review by PettitWesley — APPROVED** [review]

**Pull request merged**